manager/orchestrator/constraintenforcer/constraint_enforcer.go

```diff
@@ -115,16 +122,27 @@ func (ce *ConstraintEnforcer) rejectNoncompliantTasks(node *api.Node) {
 		// Ensure that the task assigned to the node
 		// still satisfies the resource limits.
 		if t.Spec.Resources != nil && t.Spec.Resources.Reservations != nil {
-			if t.Spec.Resources.Reservations.MemoryBytes > availableMemoryBytes {
+			if t.Spec.Resources.Reservations.MemoryBytes > available.MemoryBytes {
 				removeTasks[t.ID] = t
 				continue
 			}
-			if t.Spec.Resources.Reservations.NanoCPUs > availableNanoCPUs {
+			if t.Spec.Resources.Reservations.NanoCPUs > available.NanoCPUs {
 				removeTasks[t.ID] = t
 				continue
 			}
-			availableMemoryBytes -= t.Spec.Resources.Reservations.MemoryBytes
-			availableNanoCPUs -= t.Spec.Resources.Reservations.NanoCPUs
+			for _, ta := range t.AssignedGenericResources {
+				// Type change or no longer available
+				if genericresource.HasResource(ta, available.Generic) {
+					removeTasks[t.ID] = t
+					break loop
+				}
+			}
+
+			available.MemoryBytes -= t.Spec.Resources.Reservations.MemoryBytes
+			available.NanoCPUs -= t.Spec.Resources.Reservations.NanoCPUs
+
+			genericresource.ClaimResources(&available.Generic,
+				&fakeStore, t.AssignedGenericResources)
```

Comment:
>  it should let the value go negative so that it ends up correct after removeTask

That was why I switch from uint64 to int64 :)

> Both the scheduler and constraint enforcer could call RemoveNodeResources to subtract the available resources used by each task. If there were insufficient resources, RemoveNodeResources would return an error

RemoveNodeResources will remove the resources. However, if you have an error you need to recover the resources you just removed.

The patern goes from:
```
for t in tasks:
    nodeResCopy = nodeRes.Copy()
    for r in res:
      if remove(r, nodeRes):
         nodeRes = nodeResCopy
         removeTask(t)
         continue taskLoop
```

Which I feel isn't a lot better than:
```
for t in tasks:

    for r in res:
      if hasResource(r, nodeRes):
         removeTask(t)
         continue taskLoop

    removeResources(res)
```